src/algorithms/contractions/ctmrg_contractions.jl

```diff
@@ -578,59 +878,283 @@ function full_infinite_environment(
+    |quadrant4|    |quadrant3|
+    |~~~~~~~~~| -- |~~~~~~~~~|
+```
 
-Apply projectors to each side of a quadrant.
+In the same manner two halfs can be used to contract the full-infinite environment.
 
 ```
-    |~~~~~~~~| -- |~~~~~~|
-    |quadrant|    |P_left| --
-    |~~~~~~~~| == |~~~~~~|
-     |    ||
-    [P_right]
-        |
+    |~~~~~~~~~~~~~~~~~~~~~~~~|
+    |         half1          |
+    |~~~~~~~~~~~~~~~~~~~~~~~~|
+      |    |         |     |
+                     |     |
+      |    |         |     |
+    |~~~~~~~~~~~~~~~~~~~~~~~~|
+    |         half2          |
+    |~~~~~~~~~~~~~~~~~~~~~~~~|
 ```
-"""
-function renormalize_corner(quadrant::AbstractTensorMap{S,3,3}, P_left, P_right) where {S}
-    return @autoopt @tensor corner[χ_in; χ_out] :=
-        P_right[χ_in; χ1 D1 D2] * quadrant[χ1 D1 D2; χ2 D3 D4] * P_left[χ2 D3 D4; χ_out]
-end
 
-"""
-    renormalize_northwest_corner((row, col), enlarged_envs::CTMRGEnv, P_left, P_right)
-    renormalize_northwest_corner(quadrant::AbstractTensorMap{S,3,3}, P_left, P_right) where {S}
-    renormalize_northwest_corner(E_west, C_northwest, E_north, P_left, P_right, ket::PEPSTensor, bra::PEPSTensor=ket)
+The environment can also be contracted directly from all its constituent tensors.
 
-Apply `renormalize_corner` to the enlarged northwest corner.
-Alternatively, provide the constituent tensors and perform the complete contraction.
+```
+    C_1 --  E_2       --  E_3       -- C_2
+     |       |             |            | 
+    E_1 -- partfunc_1 -- partfunc_2 -- E_4
+     |       |             |            |
+                           |            |
+     |       |             |            |
+    E_8 -- partfunc_4 -- partfunc_3 -- E_5
+     |       |             |            |
+    C_4 --  E_7       --  E_6       -- C_3
+```
+
+Alternatively, contract the environment with a vector `x` acting on it
 
 ```
-    C_northwest -- E_north -- |~~~~~~|
-         |           ||       |P_left| --
-      E_west=   == ket-bra == |~~~~~~|
-         |           ||
-      [~~~~~P_right~~~~~]
-               |
+    C_1 --    E_2     --    E_3      --  C_2
+     |         |             |            | 
+    E_1 -- partfunc_1 -- partfunc_2  --  E_4
+     |         |             |            |
+                             |            |
+                             |            |
+    [~~~~~x~~~~~~]           |            |
+     |         |             |            |
+    E_8 -- partfunc_4 -- partfunc_3  --  E_5
+     |         |             |            |
+    C_4 --    E_7     --    E_6      --  C_3
+
 ```
+
+or contract the adjoint environment with `x`, e.g. as needed for iterative solvers.
 """
-function renormalize_northwest_corner((row, col), enlarged_envs, P_left, P_right)
-    return renormalize_northwest_corner(
-        enlarged_envs[NORTHWEST, row, col],
-        P_left[NORTH, row, col],
-        P_right[WEST, _next(row, end), col],
+function full_infinite_environment(
```

Comment:
ensure `fullinfinite_...` or `full_infinite_...` is consistent with `halfinfinite_...`